File: 1_PythonServer/app/schemas/global_memory.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class GlobalMemoryValueResponse(BaseModel):
    content: str
    keywords: list[str] = Field(default_factory=list)


class GlobalMemoryEventResponse(BaseModel):
    event_index: int
    operation: Literal["add", "update", "delete"]
    memory_id: str
    source: str
    created_at: str
    reason: str
    before: GlobalMemoryValueResponse | None = None
    after: GlobalMemoryValueResponse | None = None


class GlobalMemoryRecordResponse(BaseModel):
    id: str
    scope: Literal["global"]
    status: Literal["active", "deleted"]
    content: str
    keywords: list[str] = Field(default_factory=list)
    created_at: str
    updated_at: str
    deleted_at: str = ""
    source: str
    last_operation: Literal["add", "update", "delete"]
    event_count: int
    events: list[GlobalMemoryEventResponse] = Field(default_factory=list)


class GlobalMemoryRecordListResponse(BaseModel):
    scope: Literal["global"]
    status: Literal["active", "deleted", "all"]
    count: int
    total_count: int
    page: int
    page_size: int
    total_pages: int
    has_previous: bool = False
    has_next: bool = False
    items: list[GlobalMemoryRecordResponse] = Field(default_factory=list)


class GlobalMemoryEventListResponse(BaseModel):
    scope: Literal["global"]
    count: int
    total_count: int
    page: int
    page_size: int
    total_pages: int
    has_previous: bool = False
    has_next: bool = False
    items: list[GlobalMemoryEventResponse] = Field(default_factory=list)


class GlobalMemoryOperationRequest(BaseModel):
    operation: Literal["add", "update", "delete"]
    memory_id: str | None = None
    content: str | None = None
    keywords: list[str] | None = None
    reason: str = Field(min_length=1)


class GlobalMemoryOperationResponse(BaseModel):
    operation: Literal["add", "update", "delete"]
    memory_id: str
    memory: GlobalMemoryRecordResponse

File: 2_ReactWeb/src/entities/global-memory/model/globalMemory.ts
export type GlobalMemoryRecordStatus = "active" | "deleted";
export type GlobalMemoryRecordView = GlobalMemoryRecordStatus | "all";
export type GlobalMemoryOperation = "add" | "update" | "delete";

export type GlobalMemoryValue = {
  content: string;
  keywords: string[];
};

export type GlobalMemoryEvent = {
  event_index: number;
  operation: GlobalMemoryOperation;
  memory_id: string;
  source: string;
  created_at: string;
  reason: string;
  before: GlobalMemoryValue | null;
  after: GlobalMemoryValue | null;
};

export type GlobalMemoryRecord = {
  id: string;
  scope: "global";
  status: GlobalMemoryRecordStatus;
  content: string;
  keywords: string[];
  created_at: string;
  updated_at: string;
  deleted_at: string;
  source: string;
  last_operation: GlobalMemoryOperation;
  event_count: number;
  events: GlobalMemoryEvent[];
};

export type GlobalMemoryRecordListResponse = {
  scope: "global";
  status: GlobalMemoryRecordView;
  count: number;
  total_count: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_previous: boolean;
  has_next: boolean;
  items: GlobalMemoryRecord[];
};

export type GlobalMemoryEventListResponse = {
  scope: "global";
  count: number;
  total_count: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_previous: boolean;
  has_next: boolean;
  items: GlobalMemoryEvent[];
};

export type GlobalMemoryOperationInput = {
  operation: GlobalMemoryOperation;
  memory_id?: string;
  content?: string;
  keywords?: string[];
  reason: string;
};

export type GlobalMemoryOperationResponse = {
  operation: GlobalMemoryOperation;
  memory_id: string;
  memory: GlobalMemoryRecord;
};
